### ISSUE TITLE
Can't launch snipe-it in docker through docker-compose

### ISSUE THREAD

**Issue by chiva** [snipe/snipe-it#4657]
#### Expected Behavior (or desired behavior if a feature request)
Create a docker container for snipeit and mysql from a docker-compose.yaml file: https://pastebin.com/1tKQT2hq

Setup pages shows, checks are OK and installation can be done correctly.

-----

#### Actual Behavior

Setup welcome page fails to show due to exception

-----

#### Please confirm you have done the following before posting your bug report:

- [X ] I have enabled debug mode 
- [X ] I have read [checked the Common Issues page](https://snipe-it.readme.io/docs/common-issues)

-----
#### Provide answers to these questions:

- Is this a fresh install or an upgrade? 
Fresh install
- Version of Snipe-IT you're running
Lastest from docker
- Version of PHP you're running
From docker container
- Version of MySQL/MariaDB you're running
5.6
- What OS and web server you're running Snipe-IT on
Ubuntu Server 16.04
- What method you used to install Snipe-IT (install.sh, manual installation, docker, etc)
Docker
- WITH DEBUG TURNED ON, if you're getting an error in your browser, include that error

- What specific Snipe-IT page you're on, and what specific element you're interacting with to trigger the error
Setup welcome page -> http://SERVER/setup
- If a stacktrace is provided in the error, include that too.
https://pastebin.com/839XCpzV
- Any errors that appear in your browser's error console.
- Confirm whether the error is reproducible on the demo: https://snipeitapp.com/demo.
No
- Include any additional information you can find in `app/storage/logs` and your webserver's logs.
- Include what you've done so far in the installation, and if you got any error messages along the way.
Can't even get to show the setup welcome/check page, as it can't connect to the database container.

There is a query exception, that comes from a PDOException. The environment variable that sets the mysql server address seems to be blank (checked inside the container).

> (1/4)                                         PDOException                                                                          PDO::__construct(): php_network_getaddresses: getaddrinfo failed: Name or service not known
> --
> in PDOConnection.php (line 43)
> at PDO->__construct('mysql:host=${MYSQL_PORT_3306_TCP_ADDR};port=3306;dbname=snipeit', 'snipeit', 'Ari.snipeit.snipeit', array(0, 2, 0, false, false))in PDOConnection.php (line 43)

I've tried to set the variable to the container name at the docker-compose.yaml file: https://pastebin.com/h0HCV9da
Then setup check page works, but when continuing to the next page (migration), there is a new exception.
https://pastebin.com/Tu9WcHam

If I refresh the page, the migration page shows correctly, and tells there is no data to migrate (which is ok in this case). From there I can finish the setup correctly and use snipeit as expected.

TL:DR: If I launch Snipe-IT in docker with the command lines from the documentation it works correctly, but when converting it to a docker-compose.yaml file, setup fails to start as the enviroment variable for the mysql server address is not set.

- Indicate whether or not you've manually edited any data directly in the database
No

Please do not post an issue without answering the related questions above. If you have opened a different issue and already answered these questions, answer them again, once for every ticket. It will be next to impossible for us to help you.

https://snipe-it.readme.io/docs/getting-help

**Comment by NobleD5** [comment]
Got working stack on ubuntu-server:17.10 host. 

That's content of my docker-compose.yml  
```yaml
# Snipe-IT container
  snipe_it:
    container_name: snipe_it
    image: snipe/snipe-it:v4.1.9
    depends_on:
      - snipe_mysql
    restart: on-failure:5
    ports:
      - "3030:80"
    networks:
      br0:
        ipv4_address: 172.18.1.8
    env_file:
      - snipe.env
    healthcheck:
      test: 'stat /etc/passwd || exit 1'
      interval: 1m
      timeout: 1m
      retries: 5
    stop_grace_period: 1m
    volumes:
      - type: bind
        source: /home/engineer/snipe_it_data
        target: /var/lib/snipeit
# ------------------------------------------------------------------------------
  # MySQL for Snipe-IT container
  snipe_mysql:
    container_name: snipe_mysql
    image: mysql:5.6
    restart: on-failure:5
    expose:
      - "3306"
    networks:
      br0:
        ipv4_address: 172.18.1.9
    env_file:
      - snipe.env
    healthcheck:
      test: 'stat /etc/passwd || exit 1'
      interval: 1m
      timeout: 1m
      retries: 5
    stop_grace_period: 2m
    volumes:
      - type: bind
        source: /home/engineer/snipe_mysql_data
        target: /var/lib/mysql
```
And here some content of snipe.env
```
MYSQL_PORT_3306_TCP_ADDR=172.18.1.9

# Snipe-IT Settings
APP_ENV=production
APP_DEBUG=true
APP_KEY=base64:blah-blah
APP_URL=http://172.18.1.8:80
APP_TIMEZONE=Europe/Moscow
APP_LOCALE=ru
```
Hope it helps.

**Comment by chiva** [comment]
Thank you for your support @NobleD5 
I see that you force the MySQL container address at the env file as I had to do: MYSQL_PORT_3306_TCP_ADDR=172.18.1.9
But with same results, exception at the migration page first time I visit it, and works correctly when refreshed.

If I remove it, I get the same results with your yml file, exception when getting to the setup page.
I don't mind adding that env variable in the file, but was curious why it worked from the command line and not from docker-compose.

**Comment by chiva** [comment]
From what I have learned in this month, `docker run` creates a bridge netowrk automatically when executed, but docker-compose since version 2 it does not, so it can't get to the mysql container.

To overcome this new behaviour, you can force the IP or set the container name as an environment variable, but it does not scale from 1 container.

To allow the same behaviour as in the `docker run` command, you have to create a bridge network between the containers and assign a alias to each container so they can be found by each other with their expected hostnames.

**A fully working docker compose file in v2 can found below:**
```yaml
version: '2.2'

services:
  # MySQL for Snipe-IT container
  snipeit-mysql:
    image: mysql:5.6
    restart: unless-stopped
    env_file:
      - .env_db_mysql
      - .env_srv
    healthcheck:
      test: 'stat /etc/passwd || exit 1'
      interval: 1m
      timeout: 1m
      retries: 5
    stop_grace_period: 2m
    volumes:
      - ./snipe_env/var/lib/mysql:/var/lib/mysql:rw
    networks:
      snipeit_net:
        aliases:
          - snipeit-mysql

# ------------------------------------------------------------------------------

  # Snipe-IT container
  snipeit-server:
    image: snipe/snipe-it:latest
    depends_on:
      - snipeit-mysql
    restart: unless-stopped
    ports:
      - "8083:80"
    env_file:
      - .env_db_mysql
      - .env_srv
    healthcheck:
      test: 'stat /etc/passwd || exit 1'
      interval: 1m
      timeout: 1m
      retries: 5
    stop_grace_period: 2m
    volumes:
      - ./snipe_env/var/lib/snipeit:/var/lib/snipeit:rw
    networks:
      snipeit_net:
        aliases:
          - snipeit-server

networks:
  snipeit_net:
    driver: bridge
```

I hope @snipe finds it useful and adds it to the documentation, as its currently missing a docker-compose file.

**Comment by timby** [comment]
@chiva Thanks for posting your solution. I tried something very similar but unfortunately I'm still experiencing the issue. Would you mind posting samples of your env_files? I feel like I'm almost there and am just missing a little piece...

**Comment by timby** [comment]
I got it up and running (I had a small, hard to detect typo). Here's my docker-compose.yml, with inline environment variables (env_file can be used instead, if desired). To keep it short, I have omitted health checks, etc.

```yml
version: "3.7"
services:
  mysql:
    image: mysql:5.6
    container_name: snipeit-mysql
    restart: always
    volumes:
      - db:/var/lib/mysql
    environment:
      - MYSQL_ROOT_PASSWORD=YOUR_SUPER_SECRET_PASSWORD
      - MYSQL_DATABASE=snipeit
      - MYSQL_USER=snipeit
      - MYSQL_PASSWORD=YOUR_snipeit_USER_PASSWORD
  snipeit:
    image: snipe/snipe-it:v4.6.18
    container_name: snipe-it
    restart: always
    depends_on:
      - mysql
    volumes:
      - data:/var/lib/snipeit
    environment:
      - APP_ENV=production
      - APP_DEBUG=false
      - APP_KEY=YOUR_APP_KEY
      - APP_URL=http://127.0.0.1:80
      - APP_TIMEZONE=America/Toronto
      - APP_LOCALE=en
      - MYSQL_PORT_3306_TCP_ADDR=mysql
      - MYSQL_DATABASE=snipeit
      - MYSQL_USER=snipeit
      - MYSQL_PASSWORD=YOUR_snipeit_USER_PASSWORD
      - MAIL_PORT_587_TCP_ADDR=smtp.whatever.com
      - MAIL_PORT_587_TCP_PORT=587
      - MAIL_ENV_FROM_ADDR=youremail@yourdomain.com
      - MAIL_ENV_FROM_NAME=Your Full Email Name
      - MAIL_ENV_ENCRYPTION=tcp
      - MAIL_ENV_USERNAME=your_email_username
      - MAIL_ENV_PASSWORD=your_email_password
    ports:
      - "80:80"
volumes:
  data: {}
  db: {}
```
Please note that this pulls down specific images, please adjust accordingly. You'll notice that I'm not defining the network and driver, the default is bridged. No alias is needed, just refer to the service name where needed (MYSQL_PORT_3306_TCP_ADDR=mysql). Finally, I'm not explicitly exposing the mysql ports to the snipeit container. This is handled by the networking. If you want to expose the mysql ports to the host, just add the ports section for "3306:3306" (or whatever you want).

**Comment by chiva** [comment]
Hey @timby! Nice to hear you got it working.
I switched jobs and no longer have access to that file 😅
